components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java

```diff
@@ -382,34 +412,55 @@ private void preSetUserClaimOnMobileNumberUpdate(Map<String, String> claims, Use
+                claims.put(IdentityRecoveryConstants.MOBILE_NUMBERS_CLAIM,
+                        String.join(multiAttributeSeparator, updatedAllNumbersList));
             }
-            Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
-                    .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_EXISTING_MOBILE_NUM.toString());
-            invalidatePendingMobileVerification(user, userStoreManager, claims);
 
-            if (supportMultipleMobileNumbers && shouldUpdateMultiMobilesRelatedClaims) {
-                if (!updatedVerifiedNumbersList.contains(existingMobileNumber)) {
+            if (isPrimaryMobileVerified(userStoreManager, user)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("The mobile number to be updated: %s is same as the existing mobile " +
+                                    "number for user: %s in domain: %s and user store: %s. Hence an SMS OTP " +
+                                    "verification will not be triggered.", maskIfRequired(mobileNumber),
+                            maskIfRequired(username), user.getTenantDomain(), user.getUserStoreDomain()));
+                }
+                Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
+                        .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_EXISTING_MOBILE_NUM.toString());
+                invalidatePendingMobileVerification(user, userStoreManager, claims);
+
+                if (supportMultipleMobileNumbers && shouldUpdateMultiMobilesRelatedClaims &&
+                        !updatedVerifiedNumbersList.contains(existingMobileNumber)) {
                     updatedVerifiedNumbersList.add(existingMobileNumber);
                     claims.put(IdentityRecoveryConstants.VERIFIED_MOBILE_NUMBERS_CLAIM,
                             String.join(multiAttributeSeparator, updatedVerifiedNumbersList));
                 }
-                if (!updatedAllNumbersList.contains(existingMobileNumber)) {
-                    updatedAllNumbersList.add(existingMobileNumber);
-                    claims.put(IdentityRecoveryConstants.MOBILE_NUMBERS_CLAIM,
-                            String.join(multiAttributeSeparator, updatedAllNumbersList));
-                }
+                return;
             }
-            return;
+
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("The mobile number to be updated: %s is same as the existing mobile " +
+                                "number for user: %s in domain: %s and user store: %s. Yet the mobile number is not " +
+                                "verified. Hence an SMS OTP verification will be triggered.",
+                        maskIfRequired(mobileNumber), maskIfRequired(username), user.getTenantDomain(),
+                        user.getUserStoreDomain()));
+            }
+            Utils.unsetThreadLocalIsOnlyVerifiedMobileNumbersUpdated();
```

Comment:
```suggestion
            
```
If the initial request is only to update verified mobile numbers list, the recovery scenario associated to the notification being sent should be `MOBILE_VERIFICATION_ON_VERIFIED_LIST_UPDATE `. Do not change it.  Do the same in the `UserEmailverificationhandler`

Reply:
Resolved via https://github.com/wso2-extensions/identity-governance/pull/948/commits/d5ef792fbc5ca807654f313f0e92ca1368ec9194